django_s3_express_cache/middleware.py

```diff
@@ -0,0 +1,270 @@
+
+    def __init__(
+        self, get_response, cache_timeout=None, page_timeout=None, **kwargs
+    ):
+        # Extract and remove custom kwarg before calling super()
+        self.time_based_prefix = kwargs.pop("time_based_prefix", None)
+        super().__init__(get_response, cache_timeout, page_timeout, **kwargs)
+        self._is_s3_backend = isinstance(self.cache, S3ExpressCacheBackend)
+        if self._is_s3_backend and not self.time_based_prefix:
+            raise ValueError(
+                "CacheMiddlewareS3Compatible requires 'time_based_prefix' "
+                "when used with S3ExpressCacheBackend."
+            )
+
+    def process_request(self, request):
+        # If not using S3 backend, fall back to default behavior.
+        if not self._is_s3_backend:
+            return super().process_request(request)
+
+        if request.method not in ("GET", "HEAD"):
+            request._cache_update_cache = False
+            return None  # Don't bother checking the cache.
+
+        # Use the appropriate cache key generator
+        cache_key = get_cache_key_s3_compatible(
+            request,
+            self.key_prefix,
+            "GET",
+            cache=self.cache,
+            time_based_prefix=self.time_based_prefix,
+        )
+        if cache_key is None:
+            # No cache information available, need to rebuild.
+            request._cache_update_cache = True
+            return None
+
+        response = self.cache.get(cache_key)
+        # if it wasn't found and we are looking for a HEAD, try looking just for that
+        if response is None and request.method == "HEAD":
+            cache_key = get_cache_key_s3_compatible(
+                request,
+                self.key_prefix,
+                "HEAD",
+                cache=self.cache,
+                time_based_prefix=self.time_based_prefix,
+            )
+            response = self.cache.get(cache_key)
+
+        if response is None:
+            # No cache information available, need to rebuild.
+            request._cache_update_cache = True
+            return None
+
+        # Derive the age estimation of the cached response.
+        max_age_seconds = get_max_age(response)
+        expires_timestamp = parse_http_date_safe(response.get("Expires"))
+
+        if max_age_seconds is not None and expires_timestamp is not None:
+            now_timestamp = int(time.time())
+            remaining_seconds = expires_timestamp - now_timestamp
+            # Use Age: 0 if local clock got turned back.
+            response["Age"] = max(0, max_age_seconds - remaining_seconds)
+
+        # hit, return cached response
+        request._cache_update_cache = False
+        return response
+
+    def process_response(self, request, response):
+        """Store the response in cache when appropriate."""
+        # If not using S3 backend, fall back to default behavior.
+        if not self._is_s3_backend:
+            return super().process_request(request)
```

Comment:
When using the `django_s3_express_cache.decorators.cache_page decorator` in compatible (Redis backend) mode, it raises the following error:

> The view asgiref.sync.SyncToAsync.__call__ didn't return an HttpResponse object. It returned None instead.

I believe the problem is caused by this line, which may be adjusted as follows:

```suggestion
            return super().process_response(request, response)
```

Reply:
Thanks for catching this @albertisfu